modules/ml/src/main/java/org/apache/ignite/ml/util/generators/DataStreamGenerator.java

```diff
@@ -125,10 +135,54 @@ public default DataStreamGenerator blur(RandomProducer rnd) {
      * @param upstreamTransformerBuilder Upstream transformer builder.
      * @return Dataset builder.
      */
-    public default DatasetBuilder<Vector, Double> asDatasetBuilder(int datasetSize, IgniteBiPredicate<Vector, Double> filter,
+    public default DatasetBuilder<Vector, Double> asDatasetBuilder(int datasetSize,
+        IgniteBiPredicate<Vector, Double> filter,
         int partitions, UpstreamTransformerBuilder upstreamTransformerBuilder) {
 
         return new DatasetBuilderAdapter(this, datasetSize, filter, partitions, upstreamTransformerBuilder);
     }
 
+    /**
+     * Fills given cache with labeled vectors from this generator and user defined mapper from vectors to keys.
+     *
+     * @param datasetSize Rows count to put.
+     * @param cache Cache.
+     * @param keyMapper Mapping from vectors to keys.
+     * @param <K> Key type.
+     */
+    public default <K> void fillCacheWithCustomKey(int datasetSize, IgniteCache<K, LabeledVector<Double>> cache,
+        Function<LabeledVector<Double>, K> keyMapper) {
+
+        Map<K, LabeledVector<Double>> batch = new HashMap<>();
+        labeled().limit(datasetSize).forEach(vec -> {
```

Comment:
Does it make sense to make in in parallel?

Reply:
I don't know. I think benchmarks using this code will help to solve this problem.